hail/python/hail/ir/ir.py

```diff
@@ -1285,7 +1285,27 @@ def _compute_type(self, env, agg_env):
 
 _function_registry = {}
 _seeded_function_registry = {}
```

Comment:
these should probably all be `collections.defaultdict([])`, there's a few places in here where we have extra code to handle the key not being present but if we consistently treated "not present" as empty list, things would be simpler, i.e. the remove_function is just:
```
f = (param_types, ret_type)
_function_registry[name] = [b for b in _function_registry[name] if b != f]
```

Also removes an if check in `_lookup_function_return_type`.

---

not required for the PR, but for future reference, I think this is an exemplary use case for defaultdict